ssl/record/rec_layer_d1.c

```diff
@@ -136,12 +136,16 @@ int dtls_buffer_record(SSL_CONNECTION *s, TLS_RECORD *rec)
     item->data = rdata;
 
 #ifndef OPENSSL_NO_SCTP
-    /* Store bio_dgram_sctp_rcvinfo struct */
-    if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
-        (SSL_get_state(ssl) == TLS_ST_SR_FINISHED
-         || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
-        BIO_ctrl(SSL_get_rbio(ssl), BIO_CTRL_DGRAM_SCTP_GET_RCVINFO,
-                 sizeof(rdata->recordinfo), &rdata->recordinfo);
+    {
+        SSL *ssl = SSL_CONNECTION_GET_SSL(s);
+
+        /* Store bio_dgram_sctp_rcvinfo struct */
+        if (BIO_dgram_is_sctp(SSL_get_rbio(ssl)) &&
+            (SSL_get_state(ssl) == TLS_ST_SR_FINISHED
+             || SSL_get_state(ssl) == TLS_ST_CR_FINISHED)) {
```

Comment:
It feels wrong to me to have to get the SSL object, in order call SSL_get_state() which converts it back internally into an `SSL_CONNECTION`. Possibly we need an internal `ossl_statem_get_state()` function to return this. Or maybe I'm over thinking it. Not sure.

Reply:
The conversion to SSL object should be basically free and the conversion to SSL_CONNECTION is just a NULL check so it should be fairly cheap. But yeah, we could have ossl_statem_get_state().

Reply:
Added function